### PR TITLE
Use asynchronous DNS lookups

### PR DESCRIPTION
**Pull request by mundschenk-at**
Use asynchronous DNS lookups to speed up the domain name discovery process.

### CODE REVIEW

File: modoboa_dmarc/views.py [hunk 1]
```diff
@@ -104,17 +94,41 @@ def get_context_data(self, *args, **kwargs):
...
+
+            def get_domain_name_from_ip(ip):
+                addr = reversename.from_address(ip)
+                try:
+                    resp = dns_resolver.query(addr, "PTR")
+                    node = resp[0].target
+                    while len(node.labels) > 3:
+                        node = node.parent()
+                    return (ip, str(node))
+                except (resolver.NXDOMAIN, resolver.YXDOMAIN, resolver.NoAnswer, resolver.NoNameservers, resolver.Timeout):
```
Comment (tonioo) [inline modoboa_dmarc/views.py]: Could you please respect the 80 columns limitation?

File: modoboa_dmarc/views.py [hunk 2]
```diff
@@ -104,17 +94,41 @@ def get_context_data(self, *args, **kwargs):
...
+                    return (None, None)
+
+            ips = (r.source_ip for r in all_records)
+            with concurrent.futures.ThreadPoolExecutor(max_workers = 16) as pool:
```
Comment (tonioo) [inline modoboa_dmarc/views.py]: Same here: ``max_workers=16``

File: modoboa_dmarc/views.py [hunk 3]
```diff
@@ -104,17 +94,41 @@ def get_context_data(self, *args, **kwargs):
...
+                dns_names = {i: n for (i, n) in list(pool.map(get_domain_name_from_ip, ips))}
+
+        for record in all_records:
             stats["total"] += record.count
+            name = dns_names.get(record.source_ip, "Not resolved")
```
Comment (tonioo) [inline modoboa_dmarc/views.py]: ``"Not resolved"`` is going to be be displayed somewhere? If so, could you please mark it as translatable?

**Comment by tonioo** [comment]
@mundschenk-at I merged #15 so now there are conflicts to solve ;-)

File: modoboa_dmarc/views.py (hunk 2 above)
Comment (mundschenk-at) [inline modoboa_dmarc/views.py]: Sorry, a PHPism ;)

File: modoboa_dmarc/views.py (hunk 3 above)
Comment (mundschenk-at) [inline modoboa_dmarc/views.py]: The string is taken directly from the existing code, but I've marked it up for translation in the new patch.

**Comment by mundschenk-at** [comment]
@tonioo I've adapted the PR and added a check for `PTR` records containing invalid top-level domains. 

Not sure what the style guide says about the inline comment, but I think the line needs some explanation.

**Comment by tonioo** [comment]
Indeed, a comment is useful. The PEP says to put 2 spaces before an inline comment.

**Comment by mundschenk-at** [comment]
Fixed.

**Comment by mundschenk-at** [comment]
@tonioo While looking at the results and the code, I've some questions about the order of the entries. The `trusted` and `threats`collections are `OrdereredDict`s, but the actual order would be more or less random (namely by whatever order the database provides, I see no explicit sort order for the query set). I'm not sure what the order should be, probably alphabetical for the valid domains, followed by the unresolved IPs?

Also, you define a `stats_per_day` dictionary that's not actually used. I guess this was planned for more detailed record-by-record reports (which sometimes, though seldom, contain useful hints why a check failed)? I don't think actual per day stats would provide a lot of insights, but getting a popup/hover tooltip with detailed record information for a certain IP could be useful.

**Comment by tonioo** [comment]
@mundschenk-at you're right, those parts are useless. I don't remember why I did this :p

**Pull request merged**